### PR TITLE
optimize performance

### PR DESCRIPTION
**Pull request by SantyWang**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/2d/renderer/RenderDrawInfo.cpp [hunk 1]
```diff
@@ -31,159 +31,81 @@
...
-    _sharedBuffer = buffer;
+    CC_ASSERT(drawInfoAttrs._drawInfoType == RenderDrawInfoType::COMP || drawInfoAttrs._drawInfoType == RenderDrawInfoType::IA);
+    _compAttrs._meshBuffer = Root::getInstance()->getBatcher2D()->getMeshBuffer(drawInfoAttrs._accId, drawInfoAttrs._bufferId);
 }
 
 gfx::InputAssembler* RenderDrawInfo::requestIA(gfx::Device* device) {
-    if (_nextFreeIAHandle >= _iaPool.size()) {
+    CC_ASSERT(drawInfoAttrs._drawInfoType == RenderDrawInfoType::IA);
+    if (!_iaAttrs._iaPool) {
+        _iaAttrs._iaPool = new ccstd::vector<gfx::InputAssembler*>{0};
```
Comment (dumganhar) [inline native/cocos/2d/renderer/RenderDrawInfo.cpp]: Use ccnew

File: native/cocos/2d/renderer/RenderDrawInfo.cpp [hunk 2]
```diff
@@ -31,159 +31,81 @@
...
+    }
+    if (_iaAttrs._nextFreeIAHandle >= _iaAttrs._iaPool -> size()) {
         initIAInfo(device);
     }
-    auto* ia = _iaPool[_nextFreeIAHandle++]; // 需要 reset
+    auto* ia = _iaAttrs._iaPool->at(_iaAttrs._nextFreeIAHandle++); // 需要 reset
```
Comment (dumganhar) [inline native/cocos/2d/renderer/RenderDrawInfo.cpp]: Why use `at` instead of `[]`?

File: native/cocos/2d/renderer/RenderDrawInfo.cpp [hunk 3]
```diff
@@ -31,159 +31,81 @@
...
-    _iaPool.clear();
+    _iaAttrs._iaPool -> clear();
+    CC_SAFE_DELETE(_iaAttrs._iaPool);
 }
 
 gfx::InputAssembler* RenderDrawInfo::initIAInfo(gfx::Device* device) {
-    if (_iaPool.empty()) {
-        _iaInfo = new gfx::InputAssemblerInfo();
+    if (_iaAttrs._iaPool->empty()) {
+        _iaAttrs._iaInfo = new gfx::InputAssemblerInfo();
```
Comment (dumganhar) [inline native/cocos/2d/renderer/RenderDrawInfo.cpp]: Use ccnew instead

Comment (SantyWang) [reply native/cocos/2d/renderer/RenderDrawInfo.cpp]: OK

File: native/cocos/2d/renderer/RenderDrawInfo.cpp (hunk 2 above)
Comment (SantyWang) [reply native/cocos/2d/renderer/RenderDrawInfo.cpp]: I change it to `[]`

File: native/cocos/2d/renderer/RenderDrawInfo.cpp (hunk 1 above)
Comment (SantyWang) [reply native/cocos/2d/renderer/RenderDrawInfo.cpp]: OK

File: cocos/2d/renderer/render-draw-info.ts [hunk 4]
```diff
@@ -18,50 +37,51 @@ export enum RenderDrawInfoType {
...
-    protected _vData: Float32Array | undefined;
-    protected _iData: Uint16Array | undefined;
-    protected _vertDirty: boolean | undefined;
-    protected _vbCount: number | undefined;
-    protected _ibCount: number | undefined;
-    protected _dataHash: number | undefined;
-    protected _isMeshBuffer: boolean | undefined;
-    protected _material: Material | undefined;
+    protected _accId = -1;
+    protected _bufferId = -1;
```
Comment (dumganhar) [inline cocos/2d/renderer/render-draw-info.ts:41]: Is invalid id `-1` ? 
Currently, we set _accId to Uint16Array, if it's -1, cpp value will be 65536. Does cpp code handle the invalid value `65535` ?

File: native/cocos/2d/renderer/Batcher2d.cpp [hunk 5]
```diff
@@ -93,39 +93,47 @@ void Batcher2d::walk(Node* node, float parentOpacity) { // NOLINT(misc-no-recurs
...
+            float localOpacity = entity->getLocalOpacity();
+            float localColorAlpha = entity->getColorAlpha();
+            entity->setOpacity(parentOpacity * localOpacity * localColorAlpha);
+        }
+        if (entity->isEnabled()) {
+            RenderEntityType entityType = entity->getRenderEntityType();
+
+            // when filling buffers, we should distinguish common components and other complex components like middleware
+            if (entityType == RenderEntityType::STATIC) {
+                std::array<RenderDrawInfo, RenderEntity::STATIC_DRAW_INFO_CAPACITY>& drawInfos = entity->getStaticRenderDrawInfos();
```
Comment (dumganhar) [inline native/cocos/2d/renderer/Batcher2d.cpp:107]: It's really too long code.
It's better to use `const auto& drawInfos = entity->getStaticRenderDrawInfos();`

File: native/cocos/2d/renderer/RenderDrawInfo.h [hunk 6]
```diff
@@ -120,57 +218,48 @@ class RenderDrawInfo final {
...
-    gfx::Buffer* _vbGFXBuffer{nullptr};
+    float* _vbBuffer{nullptr};
     // weak reference
-    gfx::Buffer* _ibGFXBuffer{nullptr};
+    uint16_t* _ibBuffer{nullptr};
+
+    union {
+        Node* _subNode{nullptr};
+        scene::Model* _model;
+        uint8_t* _sharedBuffer;
```
Comment (minggo) [inline native/cocos/2d/renderer/RenderDrawInfo.h:258]: Default values.

Comment (SantyWang) [reply native/cocos/2d/renderer/RenderDrawInfo.h:258]: Union can only have one default value

File: native/cocos/2d/renderer/Batcher2d.cpp (hunk 5 above)
Comment (SantyWang) [reply native/cocos/2d/renderer/Batcher2d.cpp:107]: Yeah, I think so, my next job is optimize this part

File: cocos/2d/renderer/render-draw-info.ts (hunk 4 above)
Comment (SantyWang) [reply cocos/2d/renderer/render-draw-info.ts:41]: It's a trick, my next pr will optimize it

**Review by minggo — APPROVED** [review]

**Review by dumganhar — APPROVED** [review]

**Pull request merged**